### PR TITLE
Implementation of the Loop-Free Alternates policy 

### PR DESCRIPTION
**Pull request by sandervrijders**
This pull request contains the implementation of the LFA policy, needed for PRISTINE T4.3 resiliency work. Added myself as maintainer. It has been tested and it works like the default policy set in the normal case (apart from the fact that it doesn't support multicast, but this is not happening in user-space at the moment either). 

It also responds to flow up and down events. The policy also takes a module parameter to react differently to flow up events, for testing purposes. Since the alternate entries are not propagated to kernel space yet, not all cases have been tested yet. In the case of no available LFAs, the policy works. Propagation of LFA entries to kernel space is foreseen in a next pull request. 

Maintainers: @msune and myself. @vmaffione Can you have a look since I cannot ACK the policy set myself. 

Closes: #639

### CODE REVIEW

File: plugins/lfa/ps.c [hunk 1]
```diff
@@ -0,0 +1,802 @@
...
+static struct pft_entry * pfte_create_ni(address_t destination,
+					 qos_id_t qos_id)
+{ return pfte_create_gfp(GFP_ATOMIC, destination, qos_id); }
+
+/* FIXME: This thing is bogus and has to be fixed properly */
+#ifdef CONFIG_RINA_ASSERTIONS
+static bool pfte_is_ok(struct pft_entry *entry)
+{ return entry ? true : false; }
+#endif
+
```
Comment (msune) [inline plugins/lfa/ps.c]: Unecessary:

either `ASSERT(entry)` or `BUG_ON(!expr)` (@vmaffione)

File: plugins/lfa/ps.c [hunk 2]
```diff
@@ -0,0 +1,802 @@
...
+static void pfte_destroy(struct pft_entry *entry)
+{
+	struct pft_port_entry *pos, *next;
+
+	ASSERT(pfte_is_ok(entry));
+
+	list_for_each_entry_safe(pos, next, &entry->alt_ports, next) {
```
Comment (msune) [inline plugins/lfa/ps.c]: Extra {}

File: plugins/lfa/ps.c [hunk 3]
```diff
@@ -0,0 +1,802 @@
...
+		}
+	}
+	*entries = 1;
+
+	(*port_ids)[0] = entry->nhop;
+
+	return 0;
+}
+
+struct pff_ps_priv {
```
Comment (msune) [inline plugins/lfa/ps.c]: I prefer putting the data structure before functions/methods

File: plugins/lfa/ps.c [hunk 4]
```diff
@@ -0,0 +1,802 @@
...
+				      struct list_head *port_id_altlists)
+{
+	struct pft_port_entry *pos;
+	struct port_id_altlist *alt;
+	int i;
+
+	ASSERT(pfte_is_ok(entry));
+
+	alt = rkmalloc(sizeof(*alt), GFP_ATOMIC);
+	if (!alt) {
```
Comment (msune) [inline plugins/lfa/ps.c]: Extra {}. Please check it with checkpatch.pl

**Comment by msune** [comment]
Please check it with `checkpatch.pl`. I suspect there are some issues with style.

**Comment by msune** [comment]
For `MAINTAINERS`, ok but remove redundant /* from paths

**Comment by msune** [comment]
For `MAINTAINERS` and general style and org. code. ACK. Not tested.

**Comment by dstaesse** [comment]
ACK

dmesg  | grep lfa

```
[ 1704.066924] rina-ps-factory(INFO): policy-set 'lfa' published successfully
[ 1704.066928] rina-pff-lfa(INFO): PFF LFA policy set loaded successfully
[ 1704.089998] rina-rmt(DBG): RMT PSs: default, lfa
[ 2212.782799] rina-pff-lfa(DBG): Port-id 2 is down
[ 2212.782801] rina-pff-lfa(DBG): Trying to find a LFA for port-id 2
[ 2212.782805] rina-pff-lfa(DBG): Found an entry that requires a LFA
[ 2236.281447] rina-pff-lfa(DBG): Port-id 2 is up
[ 2236.281449] rina-pff-lfa(DBG): Port-id 2 went up, checking forwarding table
```

**Comment by vmaffione** [comment]
Compilation warnings:
````
make -C ../../linux M=$PWD
make[1]: Entering directory '/home/vmaffione/irati/linux'
  LD      /home/vmaffione/irati/plugins/lfa/built-in.o
  CC [M]  /home/vmaffione/irati/plugins/lfa/ps.o
/home/vmaffione/irati/plugins/lfa/ps.c:80:13: warning: ‘pft_pe_is_ok’ defined but not used [-Wunused-function]
 static bool pft_pe_is_ok(struct pft_port_entry *pe)
             ^
/home/vmaffione/irati/plugins/lfa/ps.c:122:13: warning: ‘pfte_is_ok’ defined but not used [-Wunused-function]
 static bool pfte_is_ok(struct pft_entry *entry)
             ^
  LD [M]  /home/vmaffione/irati/plugins/lfa/pff-lfa.o
  Building modules, stage 2.
  MODPOST 1 modules
  CC      /home/vmaffione/irati/plugins/lfa/pff-lfa.mod.o
  LD [M]  /home/vmaffione/irati/plugins/lfa/pff-lfa.ko
make[1]: Leaving directory '/home/vmaffione/irati/linux'

````

**Comment by vmaffione** [comment]
Tested, it works --> ACK

**Pull request merged**